runtime/derive/src/lib.rs

```diff
@@ -0,0 +1,176 @@
+
+                Self {
+                    #(#field_names3,)*
+                }
+            }
+        }
+    })
+}
+
+// Example input:
+// #[repr(u32)]
+// #[derive(AscType)]
+// enum JsonValueKind {
+//     Null,
+//     Bool,
+//     Number,
+//     String,
+//     Array,
+//     Object,
+// }
+//
+// Example output:
+// impl AscType for JsonValueKind {
+//     fn to_asc_bytes(&self) -> Vec<u8> {
+//         let discriminant: u32 = match *self {
+//             JsonValueKind::Null => 0u32,
+//             JsonValueKind::Bool => 1u32,
+//             JsonValueKind::Number => 2u32,
+//             JsonValueKind::String => 3u32,
+//             JsonValueKind::Array => 4u32,
+//             JsonValueKind::Object => 5u32,
+//         };
+//         discriminant.to_asc_bytes()
+//     }
+//
+//     fn from_asc_bytes(asc_obj: &[u8]) -> Self {
+//         let mut u32_bytes: [u8; size_of::<u32>()] = [0; size_of::<u32>()];
+//         u32_bytes.copy_from_slice(&asc_obj);
+//         let discr = u32::from_le_bytes(u32_bytes);
+//         match discr {
+//             0u32 => JsonValueKind::Null,
+//             1u32 => JsonValueKind::Bool,
+//             2u32 => JsonValueKind::Number,
+//             3u32 => JsonValueKind::String,
+//             4u32 => JsonValueKind::Array,
+//             5u32 => JsonValueKind::Object,
+//             _ => panic!("value {} is out of range for {}", discr, "JsonValueKind"),
+//         }
+//     }
+// }
+fn asc_type_derive_enum(item_enum: ItemEnum) -> TokenStream {
+    let enum_name = &item_enum.ident;
+    let enum_name_iter = std::iter::repeat(enum_name);
+    let enum_name_iter2 = enum_name_iter.clone();
+    let (impl_generics, ty_generics, where_clause) = item_enum.generics.split_for_impl();
+    let variant_paths: Vec<_> = item_enum
+        .variants
+        .iter()
+        .map(|v| {
+            assert!(v.discriminant.is_none());
+            &v.ident
+        })
+        .collect();
+    let variant_paths2 = variant_paths.clone();
+    let variant_discriminant = 0..(variant_paths.len() as u32);
+    let variant_discriminant2 = variant_discriminant.clone();
+
+    TokenStream::from(quote! {
+        impl#impl_generics AscType for #enum_name#ty_generics #where_clause {
+            fn to_asc_bytes(&self) -> Vec<u8> {
+                let discriminant: u32 = match *self {
+                    #(#enum_name_iter::#variant_paths => #variant_discriminant,)*
+                };
+                discriminant.to_asc_bytes()
+            }
+
+            fn from_asc_bytes(asc_obj: &[u8]) -> Self {
+                let mut u32_bytes: [u8; size_of::<u32>()] = [0; size_of::<u32>()];
+                u32_bytes.copy_from_slice(&asc_obj);
+                let discr = u32::from_le_bytes(u32_bytes);
+                match discr {
+                    #(#variant_discriminant2 => #enum_name_iter2::#variant_paths2,)*
+                    _ => panic!("value {} is out of range for {}", discr, stringify!(#enum_name))
```

Comment:
This is the critical difference from what we did before, we now catch if the value is outside the range of valid values for an enum.